org/mozilla/jss/ssl/javax/JSSSocketChannel.java

```diff
@@ -242,53 +240,54 @@ public synchronized long read(ByteBuffer[] dsts, int offset, int length) throws
             return -1;
         }
 
-        readBuffer.clear();
-
-        long remoteRead = 0;
         long unwrapped = 0;
         long decrypted = 0;
 
         try {
+            SSLEngineResult result;
             do {
-                int buffer_size = boundRead(readBuffer.remaining());
-                ByteBuffer src = ByteBuffer.wrap(readBuffer.array(), readBuffer.position(), buffer_size);
-
-                int this_read = 0;
-                if (consumed != null) {
-                    this_read += consumedChannel.read(src);
-                } else {
-                    this_read += readChannel.read(src);
-                }
-
-                readBuffer.position(readBuffer.position() + this_read);
-                remoteRead += this_read;
-
-                if (remoteRead == 0) {
-                    return 0;
+                int n = remoteRead();
+
+                if (readBuffer.position() == 0) {
+                    // We didn't read anything and there is no left-over data.
+                    // If handshake already completed, we can continue because
+                    // there may be data in the unwrapped buffer FD that has yet
+                    // to be delivered to the application.
+                    //
+                    // But if we did not finish the handshake and there were no
+                    // new data, we have to return here.  Calling unwrap() with
+                    // no data, prior to handshake completion, breaks something.
+                    //
+                    // Note we cannot call engine.getHandshakeStatus() to perform
+                    // this check; it is NOT a passive routine and reading it
+                    // early actually causes the same breakage.
+                    //
+                    if (!handshakeCompleted) {
+                        return decrypted > 0 ? decrypted : n;
```

Comment:
I'm not sure I agree with this.

When `decrypted > 0`, we obviously return `decrypted`.

However, in the case where `decrypted <= 0` (really, it must be 0), the contract of our `read()` call is the number of bytes decrypted, which is 0. It isn't the number of bytes read from the remote peer.

The best example of this is during handshake: `read` during handshake returns 0 if no data is present -- not because no wire data was transferred (there presumably might have been) -- but because none of it resulted in any decryptable data. (Ignoring 0RTT). 

IMO, since `decrypted < 0` won't ever occur, it should be safe to return `decrypted`.

Reply:
The goal here is to catch the case where `remoteRead() < 0` indicating end of stream.  I think this can occur if the connection gets terminated abnormally, i.e. the TLS termination alerts didn't get sent/processed (which needs to happen so `inboundClosed` gets set).

If all streams get closed "properly" then I agree with your proposal.  But otherwise, I want to be very careful to avoid situations where `JSSSocketChannel.read()` could return `0` indefinitely.

What do you think about this scenario @cipherboy?  Is there a better way to handle it?

Reply:
If `remoteRead` returns non-zero, shouldn't we just return `remoteRead`'s value? When the underlying TCP connections terminate unexpectedly, IMO we should never attempt to wrap/unwrap, even if we had old or uncompacted data lying around. We should probably close the TLS SSLEngine's inbound and exit with `-1`, regardless of handshake status. I think then this entire nested if statement gets simplified and removed, replaced with one that always checks the `remoteRead` value. 

This might require a little restructuring, as I think the `close()` handler doesn't yet understand that the inbound was already closed and to only close the outbound in that case (or visa-versa) but IMO it is a worthwhile improvement.

Am I missing a case where this wouldn't be correct? I thought it was one of the core guarantees from TLS's RFC.

Reply:
> However, in the case where decrypted <= 0 (really, it must be 0), the contract of our read() call is the number of bytes decrypted, which is 0. It isn't the number of bytes read from the remote peer.

> If remoteRead returns non-zero, shouldn't we just return remoteRead's value?

The return contract of `SocketChannel.read()` is number of bytes read (possibly `0`) or `-1` when *this channel* reaches end of stream.  If `remoteRead()` returns `-1`, there might still be data in `readBuffer` or in the engine's internal read buffer, so we can't return `-1` until `unwrap` yields no more data.

> IMO we should never attempt to wrap/unwrap, even if we had old or uncompacted data lying around.

Disagree.  Because there may be undecrypted data sitting in buffers waiting to be processed, we *cannot even know* if the closure of the parent socket's input stream is expected or abnormal.  Therefore even when `remoteRead()` returns `-1`, we have to wait until `unwrap` yields no more data before cleaning up and returning `-1` from `read()`.

> This might require a little restructuring

I'm sure the structure can be improved.  But... have I convinced you that the statements / expressions we're arguing about are at least not *obviously* wrong or harmful?  If so, could we merge this PR, add a regression test adapted from my test program, and go forward from there?

Otherwise... could you write a program to demonstrate bug(s) in the code fragments we're arguing about.  I suggest it not because I'm sure my code is correct (I am not!), but because concrete demonstrations of issues provided a basis for fixing them (and then for regression tests).

tl;dr this PR fixes some major issues (proven by the test program); there may well be problems with my approach but let's either make them concrete so I can understand and fix them, or merge this PR.

Reply:
Ah I'm with you now.

I keep thinking -1 return value is only on an _error_ -- i.e., stream _incorrectly_ closed or was otherwise terminated.

Since it returns -1 on mundane "no more data for now" things, yeah, I think this is a good. I'll go ahead and merge.

Reply:
@cipherboy small correction: -1 means (or is supposed to mean) "no more data *ever*", which may or may not be mundane :)